include/boost/config/compiler/codegear.hpp

```diff
@@ -9,6 +9,139 @@
+//
+#if defined(_MSC_VER) && (_MSC_VER <= 1200)
+#  define BOOST_NO_ARGUMENT_DEPENDENT_LOOKUP
+#  define BOOST_NO_VOID_RETURNS
+#endif
+//
+
+*/
+
+// Specific settings for Embarcadero drivers
+#  define BOOST_EMBTC          __CODEGEARC__
+#  define BOOST_EMBTC_FULL_VER ((__clang_major__      << 16) | \
+                                (__clang_minor__      <<  8) | \
+                                 __clang_patchlevel__         )
+
+// Detecting which Embarcadero driver is being used
+#if defined(BOOST_EMBTC)
+#  if defined(_WIN64)
+#    define BOOST_EMBTC_WIN64
+#    define BOOST_EMBTC_WINDOWS
+#    ifndef BOOST_USE_WINDOWS_H
+#      define BOOST_USE_WINDOWS_H
+#    endif
+#  elif defined(_WIN32)
+#    define BOOST_EMBTC_WIN32C
+#    define BOOST_EMBTC_WINDOWS
+#    ifndef BOOST_USE_WINDOWS_H
+#      define BOOST_USE_WINDOWS_H
+#    endif
+#  elif defined(__APPLE__) && defined(__arm__)
+#    define BOOST_EMBTC_IOSARM
+#    define BOOST_EMBTC_IOS
+#  elif defined(__APPLE__) && defined(__aarch64__)
+#    define BOOST_EMBTC_IOSARM64
+#    define BOOST_EMBTC_IOS
+#  elif defined(__ANDROID__) && defined(__arm__)
+#    define BOOST_EMBTC_AARM
+#    define BOOST_EMBTC_ANDROID
+#  elif
+#    if defined(BOOST_ASSERT_CONFIG)
+#       error "Unknown Embarcadero driver"
+#    else
+#       warning "Unknown Embarcadero driver"
+#    endif /* defined(BOOST_ASSERT_CONFIG) */
+#  endif
+#endif /* defined(BOOST_EMBTC) */
+
+#if defined(BOOST_EMBTC_WINDOWS)
+   // About TR1 headers and features:
+   // BOOST_HAS_TR1_*-style definitions are placed in "boost\tr1\detail\config.hpp",
+   // because there, they will always be found, but here, they may be not check.
+
+   //   // This is needed to allow 64-bit integers in cstdint.hpp
+   //#  define BOOST_HAS_MS_INT64
+
+   //#  define BOOST_HAS_RVALUE_REFS
+
+   //   // Boost.Fusion should not use its preprocessed templates,
+   //   // due to we can use its variadic cpp11 templates. If that
+   //   // macro is not defined it tries to use both at the same time
+   //   // giving a redefinition error.
+   //#  define BOOST_FUSION_DONT_USE_PREPROCESSED_FILES
+
+   //   // Bcc64 don't have this <fenv.h> header
+   //#  define BOOST_NO_FENV_H
+
+   //   // This is not necessary with the last versions of bcc64
+   //#  if !defined(__FUNC__)
+   //#    define __FUNC__ __func__
+   //#  endif
+
+   // FVTODO: This was uncommended with 1_55 - prolly worth testing this by uncommenting it.
+   // Dinkumware on Win32 and Win64 platforms has <cstdint>
+   // #  define BOOST_HAS_STDINT_H
+#endif /* BOOST_EMBTC_WINDOWS */
+
+#  undef BOOST_COMPILER
+#  define BOOST_COMPILER "Embarcadero-Clang C++ version " BOOST_STRINGIZE(__CODEGEARC__) " clang: " __clang_version__
+// #  define __CODEGEARC_CLANG__ __CODEGEARC__
+// #  define __EMBARCADERO_CLANG__ __CODEGEARC__
+// #  define __BORLANDC_CLANG__ __BORLANDC__
+
+#else // #if !defined(__clang__)
+
+# define BOOST_CODEGEARC  __CODEGEARC__
+# define BOOST_BORLANDC   __BORLANDC__
```

Comment:
This breaks the whole point of `BOOST_compiler` macros that should tell if it is really the particular compiler and not an other that mimics it.

Reply:
It does not. The definition is only made when __CODEGEARC__ is originally defined, meaning the Borland/Codegear/Embarcadero C++ compilers, and when __clang__ is not defined, meaning that it is not the Embarcadero C++ clang-based compiler(s). There is no other compiler that defines __BORLANDC__ and __CODEGEARC__ anywhere.

Reply:
It does, you are defining both simultaneously.

Reply:
What is the difference. Both \_\_BORLANDC\_\_ and \_\_CODEGEARC\_\_ are always defined to the same value whenever \_\_CODEGEARC\_\_ is defined. I want both BOOST_BORLANDC and BOOST_CODEGEARC defined whenever the non-clang-based compilers are detected.

Reply:
There is no point in having both when they mean the same thing, it only produces a confusion

Reply:
Both are needed because there is code in other libraries referrring to the old bcc32 compiler using both \_\_BORLANDC\_\_ and \_\_CODEGEARC\_\_.

Reply:
@jzmaddock is the intention that there should only ever be one BOOST_COMPILER macro defined?

Reply:
>is the intention that there should only ever be one BOOST_COMPILER macro defined?

Yes but that's not really the mistake here: Borland and Codegear are the same compiler give or take a few updates, we should probably have kept the same header and name.

Personally I think it's fine to define both under these circumstances.  Note that with the Sun->Oracle name change we've basically just ignored the issue and continued calling it "sunpro" and no one seems to have complained (yet).